domains/client/domain-operator/src/operator.rs

```diff
@@ -178,19 +191,167 @@ where
+                        {
+                            trace!(%block_number, "Acknowledged block import from consensus chain.");
+                            if acknowledgement_sender.send(()).await.is_err() {
+                                return Err(sp_consensus::Error::Other(
+                                    format!("Can't acknowledge block import #{}", block_number)
+                                        .into(),
+                                ));
+                            }
+
+                            if block_number >= target_block_number.into() {
+                                break;
+                            }
+                        }
+
+                        // Drain Substrate block imported notifications
+                        let mut imported_block_notification_stream =
+                            Box::pin(params.operator_streams.imported_block_notification_stream);
+
+                        while let Some(import_notification) =
+                            imported_block_notification_stream.next().await
+                        {
+                            let block_number = *import_notification.header.number();
+                            trace!(%block_number, "Block imported from consensus chain.");
+
+                            if block_number >= target_block_number.into() {
+                                break;
+                            }
+                        }
+
+                        // Restore parameters
+                        unsafe {
+                            params.operator_streams.block_importing_notification_stream =
+                                Box::<IBNS>::into_inner(Pin::into_inner_unchecked(
+                                    block_importing_notification_stream,
+                                ));
+                            params.operator_streams.imported_block_notification_stream =
+                                Box::<CIBNS>::into_inner(Pin::into_inner_unchecked(
+                                    imported_block_notification_stream,
+                                ));
+                        }
+                    }
+
+                    crate::domain_worker::start_worker(
+                        spawn_essential.clone(),
+                        consensus_client.clone(),
+                        params.consensus_offchain_tx_pool_factory.clone(),
+                        params.maybe_operator_id,
+                        bundle_producer,
+                        bundle_processor.clone(),
+                        params.operator_streams,
+                    )
+                    .boxed()
+                };
+
+                spawn_essential.spawn_essential_blocking(
+                    "domain-operator-worker",
+                    None,
+                    Box::pin(start_worker_task),
+                );
+
+                // Keep essential task running
+                pending::<()>().await;
```

Comment:
Would it be better to select on the spawned tasks finishing instead?

Reply:
Can you rephrase or suggest a pseudo-code? This line prevents essential task from exiting that will cause the process exit.

Reply:
Yes, I read the comment 🙂

I think I was confused by the nested task structure, and the complicated code here. It’s a bit simpler now.

As long as we’re spawning all our tasks as essential tasks, there’s no need to select on any of them finishing.

I also agree with Ning’s suggestion, if we can combine the starter task and the long term task, there should be no need for a `pending` here.

(Feel free to resolve this once you’ve read it.)